pytest_flask_sqlalchemy/fixtures.py

```diff
@@ -28,8 +28,11 @@ def _transaction(request, _db, mocker):
     Create a transactional context for tests to run in.
     '''
     # Start a transaction
-    connection = _db.engine.connect()
-    transaction = connection.begin()
+    try:
+        connection = _db.engine.connect()
+        transaction = connection.begin()
+    except:
+        return
```

Comment:
If we're going to go the route of exception handling, it might be useful to be more specific about the call that expect to fail, and to raise a warning in the process. What exception gets raised precisely when network services aren't available, and which call raises it? For example, if the `connect()` raises a `DatabaseError` with a string like `Database unavailable`, I think a more specific approach might be something like:

```python
    try:
        connection = _db.engine.connect()
    except sqlalchemy.exceptions.DatabaseError as e:
        if str(e) == 'Database unavailable':
            warnings.warn('Database unavailable, skipping transactional tests', RuntimeWarning)
            return
        else:
            raise e

    transaction = connection.begin()
```

Reply:
Sounds good - looking at the [sqlalchemy exception definitions](https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_2_2/lib/sqlalchemy/exc.py), we can catch a few exceptions based on their Python type, and they broadly speaking wrap existing [DB-API exceptions](https://www.python.org/dev/peps/pep-0249/#exceptions)).

From testing locally using the `pg8000` driver, the inability to connect to PostgreSQL if there isn't an instance running/available surfaces as an [`InterfaceError`](https://www.python.org/dev/peps/pep-0249/#interfaceerror).

That said there are some [current notes](https://github.com/sqlalchemy/sqlalchemy/blob/8a274e0058183cebeb37d3a5e7903209ce5e7c0e/doc/build/errors.rst#interfaceerror) in the `sqlalchemy` docs hinting that database exceptions may be driver-specific -- I'll try to check how `psycopg2` behaves soon as a comparison.

Reply:
Hrm.  There might have to be some amount of driver-specific handling in order to cater for both `pg8000` and `psycopg2` in this situation.  They surface connection errors differently:

```py
>>> psycopg2 = create_engine('postgresql+psycopg2://test@localhost/test')
>>> pg8000 = create_engine('postgresql+pg8000://test@localhost/test')
...
>>> psycopg2.connect()
...
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
...
>>> pg8000.connect()
...
sqlalchemy.exc.InterfaceError: (pg8000.exceptions.InterfaceError) Can't create a connection to host localhost and port 5432 (timeout is None and source_address is None).
...
```

Reply:
NB: Both of those exceptions inherit from `sqlalchemy`'s [`DBAPIError`](https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_2_2/lib/sqlalchemy/exc.py#L333) so that looks like it's the most-precise common exception type without requiring per-driver behaviour.  Catching that over a `connect(...)` operation seems _reasonably_ specific.

Reply:
Also tested for `py-postgresql` (found via the PostgreSQL wiki [Python](https://wiki.postgresql.org/wiki/Python) page):

```py
>>> pypg = create_engine('postgresql+pypostgresql://test@localhost/test')
...
>>> pypg.connect()
...
sqlalchemy.exc.DBAPIError: (postgresql.exceptions.Error) connection already closed
...
```

Reply:
tl;dr - `DBAPIError` seems to me like the exception to catch here, based on testing a few drivers.  Resolving this conversation thread, but feel free to re-open if it could use further investigation.

Reply:
(hrm.  I do get the sense this approach might be a slight oversimplification and that I could still be missing behavioural problems / implications for the library here.  I'll spend a bit more time considering the pull request and intent of it all)

Reply:
OK - I *think* I've managed to get back out of that moment of self-doubt and **partially** reassure myself that catching subclasses of `DBAPIError` during `sqlalchemy`'s `connect` method is a reasonable starting point to detect runtime connection errors.

The lingering doubt I have is around whether there are *other* potential sources of DB-API level failures - invalid connection strings perhaps, intermittent network conditions possibly - that could also bubble up and be caught by a `DBAPIError` handler (it's the root of the wrapped DB-API exception types, after all).

Then a question would be: is that any worse than the current behaviour? - and the answer is yes, because the specific exception would be masked by the handler and we'd simply know that 'something went wrong' but without much context to discover what.

And that puts us back, I think, to a situation where it'd be necessary to implement per-driver-specific string matching on error messages, which really doesn't seem ideal (although it would work).

Hrmm.  Requires a bit more thought (and I'd appreciate any more feedback on whether this is sounding on-track or any missing details)

Reply:
I feel this isn't an ideal solution either, but one idea that has sprung to mind is to allow this to be an opt-in behaviour via configuration.  That removes the need for any hard-to-maintain driver-specific code in the library and - at least to some degree - means that users of the library are less likely to run into unexpected test-time behaviour since the feature would have to be discovered and enabled explicitly.